airbyte-integrations/connectors/source-slack/source_slack/components/join_channels.py

```diff
@@ -0,0 +1,123 @@
+    url_base = "https://slack.com/api/"
+    http_method = "POST"
+    primary_key = "id"
+
+    def __init__(self, channel_filter: List[str] = None, **kwargs):
+        self.channel_filter = channel_filter or []
+        super().__init__(**kwargs)
+
+    def path(self, **kwargs) -> str:
+        return "conversations.join"
+
+    def parse_response(self, response: requests.Response, stream_slice: Mapping[str, Any] = None, **kwargs) -> Iterable:
+        """
+        Override to simply indicate that the specific channel was joined successfully.
+        This method should not return any data, but should return an empty iterable.
+        """
+        is_ok = response.json().get("ok", False)
+        if is_ok:
+            self.logger.info(f"Successfully joined channel: {stream_slice['channel_name']}")
+        else:
+            self.logger.info(f"Unable to joined channel: {stream_slice['channel_name']}. Reason: {response.json()}")
+        return []
+
+    def request_body_json(self, stream_slice: Mapping = None, **kwargs) -> Optional[Mapping]:
+        if stream_slice:
+            return {"channel": stream_slice.get("channel")}
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        """
+        The pagination is not applicable to this Service Stream.
+        """
+        return None
+
+
+class ChannelsRetriever(SimpleRetriever):
+    def __post_init__(self, parameters: Mapping[str, Any]):
+        super().__post_init__(parameters)
+        self.stream_slicer = SinglePartitionRouter(parameters={})
+        self.record_selector.transformations = []
+
+    def should_join_to_channel(self, config: Mapping[str, Any], record: Record) -> bool:
+        """
+        The `is_member` property indicates whether the API Bot is already assigned / joined to the channel.
+        https://api.slack.com/types/conversation#booleans
+        """
+        return config["join_channels"] and not record.get("is_member")
+
+    def make_join_channel_slice(self, channel: Mapping[str, Any]) -> Mapping[str, Any]:
+        channel_id: str = channel.get("id")
+        channel_name: str = channel.get("name")
+        LOGGER.info(f"Joining Slack Channel: `{channel_name}`")
+        return {"channel": channel_id, "channel_name": channel_name}
+
+    def join_channels_stream(self, config) -> JoinChannelsStream:
+        token = config["credentials"].get("api_token") or config["credentials"].get("access_token")
+        authenticator = TokenAuthenticator(token)
+        channel_filter = config["channel_filter"]
+        return JoinChannelsStream(authenticator=authenticator, channel_filter=channel_filter)
+
+    def join_channel(self, config: Mapping[str, Any], record: Mapping[str, Any]):
+        list(
+            self.join_channels_stream(config).read_records(
+                sync_mode=SyncMode.full_refresh,
+                stream_slice=self.make_join_channel_slice(record),
+            )
+        )
+
+    def read_records(
```

Comment:
Why "read_records" method was chosen for "join channel" injection?

Main disadvantage here is that we had to copy full content of original "read_records" instead of just calling super().read_records(....) and then adding our customization.

maybe better point of injection could be parse_response or transform  functions.

Reply:
Joining to channel logic is not transformation or parse response. We should join to channel once we receive a channel that was not joined. As it was implemented in python code. We perform all api requests in a one place and have proper handling in case of fail. As this is not logic of parse response or transformation, in feature we can remove this custom component when similar logic appear in  low-code cdk.